From a6f76c506d6e09255036cec1eb5d31cf5b6bd970 Mon Sep 17 00:00:00 2001
From: Nguyen Thanh Trung <nguyenthanh.trung@nomovok.vn>
Date: Wed, 11 Dec 2013 16:22:34 +0700
Subject: [PATCH] use urandom instead of random for retouch_binaries

---
 updater/install.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/updater/install.c b/updater/install.c
index 7b4b99b..d0baec8 100644
--- a/updater/install.c
+++ b/updater/install.c
@@ -450,7 +450,7 @@ Value* RetouchBinariesFn(const char* name, State* state,
     bool override_set = false;
     int32_t random_base = time(NULL) % 1024;
     // some more randomness from /dev/random
-    FILE *f_random = fopen("/dev/random", "rb");
+    FILE *f_random = fopen("/dev/urandom", "rb");
     uint16_t random_bits = 0;
     if (f_random != NULL) {
         fread(&random_bits, 2, 1, f_random);
-- 
1.7.9.5

